### PR TITLE
Add encoded elevation to EdgeInfo - serialize in route and trace_attributes JSON

### PR DESCRIPTION
**Pull request by dnesbitt61**
# Issue

This PR adds elevation along all edges to Valhalla tiles. Elevation is added to each node (NodeInfo) and to each edge (EdgeInfo) using delta encoding in a byte array. Adding elevation in this manner does not increase tile size by more than about 2% (so there are not tile building options - elevation is always added where available. Edges that lie outside regions where elevation is available (as specified by the directory in the config file) have a `has_elevation` flag set to false.

Elevation along an edge that is identified as a bridge or a tunnel is special: the elevation at the beginning and end of the bridge/tunnel edge is set at the end nodes and elevation along the edge is set using a linear interpolation between the elevation at the end nodes. This "fixes" elevation so it does not drop to the terrain below a bridge or rise to the terrain on top of a tunnel. However, this does not account for the true elevation along the bridge/tunnel if they rise of fall.

Elevation can be requested in route, trace_route, and trace_attribute calls by specifying elevation_interval. An elevation interval of 30 meters is recommended since this is the elevation interval used when encoding elevation in tiles.

Sample route request with elevation returned:
`http://localhost:8002/route?json={"locations":[{"lat":40.134140,"lon":-75.515294},{"lat":40.132871,"lon":-75.518312}],"costing":"bicycle","elevation_interval":30}`

JSON response includes:
```
"elevation_interval": 30.0,
        "elevation": [
              36.3,
              36.7,
              37.5,
              38.3,
              38.9,
              40.1,
              40.5,
              40.4,
              39.2,
              38.4,
              38.3,
              38.6,
              38.9,
              39.1
        ],
```

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [x] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by kevinkreiser** [comment]
awesome, great to have you back twiddling those bits!

File: proto/options.proto [hunk 1]
```diff
@@ -482,4 +482,7 @@ message Options {
     uint32 matrix_locations = 54;                                  // a one to many or many to one time distance matrix. Does not affect
   }                                                                // sources_to_targets when either sources or targets has more than 1 location
                                                                    // or when CostMatrix is the selected matrix mode.
+  oneof has_elevation_interval {
```
Comment (kevinkreiser) [inline proto/options.proto:485]: do we need to differentiate between default and not set? if not we can remove the `oneof`

Comment (dnesbitt61) [reply proto/options.proto:485]: Does this default to 0.0 in the case where elevation_interval is not included in the JSON request? If so, then I think we can remove oneof.

Comment (kevinkreiser) [reply proto/options.proto:485]: yep it does https://protobuf.dev/programming-guides/proto3/#default

Comment (dnesbitt61) [reply proto/options.proto]: pushed a fix. I kept code to make sure the value is in range when pbf inputs are used.

**Review by nilsnolde — COMMENTED** [review]
This is a great addition, thanks for the effort to push this into the public project. I don't have much to add other than a few nits. I didn't look much at stuff past triplegbuilder.cc, but until there it made all sense. Great that you handled bridges & tunnels. That's currently not possible using the /height service with a route result!

One major drawback in this PR is the test coverage IMO. Just a single route request would probably exercise the majority of changed code. We already have `test/pixels.h` which some elevation tests are using. Couldn't a gurka test be coerced to build a large-ish map in the geography of that elevation tile? Would you be up for that @dnesbitt61? Or maybe you thought about smth else to increase coverage?

File: src/mjolnir/graphtilebuilder.cc [hunk 2]
```diff
@@ -347,15 +359,17 @@ void GraphTileBuilder::StoreTileData() {
...
     header_builder_.set_edgeinfo_offset(header_builder_.complex_restriction_reverse_offset() +
                                         reverse_restriction_size);
     for (const auto& edgeinfo : edgeinfo_list_) {
       in_mem << edgeinfo;
     }
+    int64_t edge_info_size = in_mem.tellp() - current_size;
 
     // Write the names
-    header_builder_.set_textlist_offset(header_builder_.edgeinfo_offset() + edge_info_offset_);
+    header_builder_.set_textlist_offset(header_builder_.edgeinfo_offset() + edge_info_size);
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply src/mjolnir/graphtilebuilder.cc:372]: after adding EdgeInfo (perhaps with elevation) the computed edge_info_offset_ is no longer valid).

File: valhalla/midgard/elevation_encoding.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+#include <type_traits>
+#include <vector>
+
+#include <valhalla/midgard/logging.h>
+
+namespace valhalla {
+namespace midgard {
+
+// Maximum sampling interval used when encoding elevation along an edge.
+const uint32_t kMaxEdgeElevationSampling = 32;
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply valhalla/midgard/elevation_encoding.h:15]: this IS a cutoff. The sampling interval is variable along an edge such that an integer number of samples occur along the edge. The variable interval allows consistent encoding forward and reverse along an edge - we use a sampling  interval that breaks the edge into an integral number of samples (see sampling_interval method).

File: valhalla/midgard/elevation_encoding.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      if (delta < -256) {
+        LOG_WARN("Elevation delta " + std::to_string(delta) + " exceeds limit!");
+        error = true;
+      }
+      delta = -128;
+      prior -= 128;
+    } else {
+      prior = v;
+    }
+    encoding.push_back(delta);
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply valhalla/midgard/elevation_encoding.h:101]: ok

File: valhalla/midgard/elevation_encoding.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    if (elevation[i] == ELEVATION_NO_DATA_VALUE) {
+      encoding.push_back(0);
+      continue;
+    }
+
+    // Convert to fixed precision offset. Clamp to 1 byte values. It should be
+    // very rare that the limit is exceeded, but if it is we try to make up for
+    // any shortfall on subsequent postings.
+    int v = to_fixed_precision(elevation[i]);
+    int delta = v - prior;
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply valhalla/midgard/elevation_encoding.h:83]: ok

File: src/mjolnir/elevationbuilder.cc [hunk 6]
```diff
@@ -70,37 +155,49 @@ void add_elevations_to_single_tile(GraphReader& graphreader,
...
+      if (directededge.bridge() || directededge.tunnel() || directededge.use() == Use::kFerry) {
+        // Get height at beginning and end of bridge/tunnel
+        std::vector<PointLL> tmp;
+        tmp.emplace_back(resampled.front());
+        tmp.emplace_back(resampled.back());
+        auto h = sample->get_all(tmp);
+        heights[0] = h[0];
+        float endheight = h[1];
+        float dh = (endheight - heights[0]) / heights.size();
+        for (int i = 1; i < static_cast<int>(heights.size()); ++i) {
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply src/mjolnir/elevationbuilder.cc:176]: ok

File: docs/docs/api/turn-by-turn/api-reference.md [hunk 7]
```diff
@@ -320,6 +320,7 @@ For example a bus request with the result in Spanish using the OSRM (Open Source
...
 | `date_time` | This is the local date and time at the location.<ul><li>`type`<ul><li>0 - Current departure time.</li><li>1 - Specified departure time</li><li>2 - Specified arrival time. Not yet implemented for multimodal costing method.</li></li>3 - Invariant specified time. Time does not vary over the course of the path. Not implemented for multimodal or bike share routing</li></ul></li><li>`value` - the date and time is specified in ISO 8601 format (YYYY-MM-DDThh:mm) in the local time zone of departure or arrival.  For example "2016-07-03T08:06"</li></ul><br> |
+| `elevation_interval` | Elevation interval (meters) for requesting elevation along the route. Valhalla data must have been generated with elevation data. If no `elevation_interval` is specified, no elevation will be returned for the route. An elevation interval of 30 meters is recommended when elevation along the route is desired. |
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply docs/docs/api/turn-by-turn/api-reference.md]: ok.

File: src/mjolnir/edgeinfobuilder.cc [hunk 8]
```diff
@@ -80,12 +85,20 @@ void EdgeInfoBuilder::set_encoded_shape(const std::string& encoded_shape) {
   std::copy(encoded_shape.begin(), encoded_shape.end(), back_inserter(encoded_shape_));
 }
 
+// Set the encoded elevation vector.
+void EdgeInfoBuilder::set_encoded_elevation(const std::vector<int8_t>& encoded_elevation) {
+  if (!encoded_elevation.empty()) {
+    std::copy(encoded_elevation.begin(), encoded_elevation.end(), back_inserter(encoded_elevation_));
+  }
+}
```
[the remark this answers is not in the document]
Comment (dnesbitt61) [reply src/mjolnir/edgeinfobuilder.cc]: done

**Comment by dnesbitt61** [comment]
I'm not too familiar with gurka tests. Are there any unit tests that build tiles with elevation that could be used to test? I don't have much time this week, but perhaps can look more next week.

**Comment by nilsnolde** [comment]
It's a bit complex-ish.. there's a header `test/pixels.h` which has the binary representation of one elevation tile somewhere in the US. That'd first need to be saved to disk locally in the test code to a path, that path has to get set in the gurka test config as the elevation path, then build tiles with that config and do a route. Or smth like that :smile:  I can help out on that early next week if you want.

**Comment by dnesbitt61** [comment]
pixels.h adds some elevation data to create a sparse set of data within a tile. Does this elevation data run along any graph edges (hopefully 2 connected edges)? The sparse elevation data would have to fully surround the road since the skadi sample method looks at adjacent elevations when forming the elevation. If so, then we could create data with this and do trace route along the road path and check that the encoded elevation is close to matching what is expected along the path.

**Comment by kevinkreiser** [comment]
im not sure if its near anything, i made that a long time ago and forget pretty much everything about where i put it. it is probably easier to create fake data on the fly in the test by just making an srtm tile with predictable data (each row of pixels has the same elevation and it maybe increases or decreases with each row in the tile). then north south roads in a gurka test would have increasing/decreasing elevation where as east west ones would have constant elevation. thats what id do for a test.

**Comment by dnesbitt61** [comment]
Trying to use pixels.h but have some questions/issues:
1) Why are values for each index included 4 times?
2) This doesn't seem close to any real values. There are very large jumps in elevation between adjacent posts. Since the elevation along an edge is encoded using delta encoding (max change in any 30 meter increment is 16 meters) I do not believe this will work as a test.

**Comment by kevinkreiser** [comment]
@dnesbitt61 you can see how i created the data here: https://github.com/valhalla/valhalla/commit/6fb8b3d6abfeaa49cdbdd41cfb842e54e144f0ed#diff-cdf1ee9054e7a0d87f469d85364ff34bc799e7655fbcc4baf3308ae4f6c10912R10

**Comment by dnesbitt61** [comment]
so duplicates (4x per index) are due to the way you created this. This shouldn't impact the test. However, the actual elevation values cannot be real if this is from N40W077.hgt. There should be no negative values in this land area. That and the large delta's between adjacent postings (such as {4521066, 1}, {4521067, 513})

**Comment by gknisely** [comment]
> @dnesbitt61 you can see how i created the data here: [6fb8b3d#diff-cdf1ee9054e7a0d87f469d85364ff34bc799e7655fbcc4baf3308ae4f6c10912R10](https://github.com/valhalla/valhalla/commit/6fb8b3d6abfeaa49cdbdd41cfb842e54e144f0ed#diff-cdf1ee9054e7a0d87f469d85364ff34bc799e7655fbcc4baf3308ae4f6c10912R10)

@kevinkreiser what log infos are we to uncomment?  re: `uncommenting the log infos`

**Comment by gknisely** [comment]
> > @dnesbitt61 you can see how i created the data here: [6fb8b3d#diff-cdf1ee9054e7a0d87f469d85364ff34bc799e7655fbcc4baf3308ae4f6c10912R10](https://github.com/valhalla/valhalla/commit/6fb8b3d6abfeaa49cdbdd41cfb842e54e144f0ed#diff-cdf1ee9054e7a0d87f469d85364ff34bc799e7655fbcc4baf3308ae4f6c10912R10)
> 
> @kevinkreiser what log infos are we to uncomment? re: `uncommenting the log infos`

@kevinkreiser nm.  I was looking at `test/sample.cc` and not `skadi/sample.cc`

**Comment by kevinkreiser** [comment]
> There should be no negative values in this land area. That and the large delta's between adjacent postings (such as {4521066, 1}, {4521067, 513})

remember that the values are in srtm format, which is big endian. to make use of the values (as is done in sample.cc) you need to flip the endianess.  you cant look at the numbers in the pixels header and make sense of them like you're trying to do. they all go through this function to be read: https://github.com/valhalla/valhalla/blob/c30e6fbe94fb1d357fad12488d347ff92cbd0120/src/skadi/sample.cc#L42-L44

**Comment by dnesbitt61** [comment]
that is what I was missing. thanks for clarifying.

File: test/gurka/test_elevation.cc [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+      EXPECT_TRUE(elevation && elevation->IsArray());
+      std::vector<float> elevation_from_skadi;
+      for (const auto& e : elevation->GetArray()) {
+        elevation_from_skadi.push_back(std::round(e.GetFloat() * 10) / 10);
+        std::cout << std::round(e.GetFloat() * 10) / 10 << std::endl;
+      }
+
+      EXPECT_EQ(elevation_along_edges.size(), elevation_from_skadi.size());
+      for (size_t i = 0; i < elevation_along_edges.size(); ++i) {
+        EXPECT_NEAR(elevation_along_edges[i], elevation_from_skadi[i], 0.5f);
```
Comment (gknisely) [inline test/gurka/test_elevation.cc:204]: @dnesbitt61 is `0.5f` tolerance ok here?

Comment (dnesbitt61) [reply test/gurka/test_elevation.cc:204]: This is fine for now. I suspect we could find cases (esp. long edges) where the accuracy could deviate more, requiring a higher tolerance. Statistically, the delta encoding should do pretty well except the cases where slope exceeds 45 percent.

**Review by kevinkreiser — APPROVED** [review]
nice collaboration here

**Comment by kevinkreiser** [comment]
seems the test doesn't work on osx. lint build might just be a random failure

**Comment by gknisely** [comment]
> seems the test doesn't work on osx. lint build might just be a random failure

Looks like I just forgot to close the file

**Review by gknisely — APPROVED** [review]

**Comment by farfromrefug** [comment]
@dnesbitt61 very nice addition. Is this only to "query" elevation along route? Or does it influence cost for routing?
Would still be nice to have a config flag to build with it or not. 2% is not big per say, but for example when you build for mobile/embedded platform every percent gain is good.
Would you accept a PR for the config option to add elevation to EdgeInfo?

**Comment by dnesbitt61** [comment]
Yes, this returns elevation along a route for routes and map-matching. This differs slightly from the height service in that this elevation will interpolate elevations between the end points of bridges and tunnels rather than (usually) matching ground level below bridges or above tunnels.
This in NOT used for costing. We do not query EdgeInfo attributes (only directed edge attributes) within costing. The weighted grade directed edge attribute is used in costing (primarily bicycle costing) to influence routing.
I think a PR to add the config option is fine. You may want to add info to responses (route and trace_route in particular) where elevation interval is specified but no elevation is stored in the routing tiles.

**Comment by farfromrefug** [comment]
@dnesbitt61 thanks for the explanation!
actually looking at the code. Wouldnt setting `elevation_interval` to 0 be enough to disable elevation encoding in tiles?

**Comment by dnesbitt61** [comment]
elevation_interval will only impact whether elevation is computed and returned in a route/trace_route call. Elevation will be stored in tiles (mjolnir/elevation_builder.cc) if elevation data is available when building tiles. This is controlled by additional_data:elevation in the json config used when building tiles.

**Pull request merged**